### ISSUE TITLE
return a sendActivity() from a dialog

### ISSUE THREAD

**Issue by TobiahZ** [Microsoft/botbuilder-js#747]
## Versions
BotBuilder 4.2.0

## Describe the bug
In Type Script, we cannot return a step.context.sendActivity('...') due to an invalid return type.

Specifically, sendActivity returns a Promise<ResourceResponse> where a Promise<DialogTurnResult<any>> is expected.

## To Reproduce
Within a Dialog, such as a WaterfallDialog, the functions are expected to return a Promise<DialogTurnResult>. Ending the function with:
`return step.context.sendActivity('...')`
causes type errors. 

To get around this, we have to call the sendActivity and then call step.next() or step.endDialog() instead.
`
await step.context.sendActivity('...');
return step.next();
`

## Expected behavior
Multiple JavaScript examples return functions within dialogs with a message through sendActivity. However, in TypeScript, this is causes a type mismatch.

We would expect to also be able to return a dialog by sending a message in TypeScript the same way you can in JavaScript, rather than having to explicitly call step.next(), step.endDialog(), step.beginDialog() etc.

## Additional context
If there is a logical reason to not end dialogs with a sendActivity, it would make sense for the return type to not match. However, this is not documented, and should be made clear to TypeScript users that dialogs needed to return a Promise<DialogTurnResult<any>> through step.next(), step.endDialog(), step.beginDialog(), etc.

[bug]

**Comment by navzam** [comment]
I ran into the same issue in JS. Returning `Dialog.EndOfTurn` (which is what `step.prompt()` returns) seems to work, but this isn't documented and I don't know whether it will cause unexpected behaviors deeper in the dialogs library.

**Comment by justinwilaby** [comment]
The signature for the `DialogTurnResult` includes `result?: T;`. Why isn't this acceptable?
```js
const result: <Promise<ResourceResponse>> = step.context.sendActivity('...');
return {
  status: DialogTurnStatus.complete,
  result
} as DialogTurnResult<<Promise<ResourceResponse>>>;
```
This should satisfy the typing and allow the promise from `step.context.sendActivity()` to be `await`ed on at some point later

**Comment by TobiahZ** [comment]
Another example of an error in TypeScript when doing a `return step.context.sendActivity('...')`

![captureconvo](https://user-images.githubusercontent.com/4392252/52656406-249c0200-2eab-11e9-9d45-71052376363d.PNG)

Still am unable to successfully `return step.context.sendActivity('...')` in typescript

**Comment by johnataylor** [comment]
The idea is that the canonical **Waterfall** is a chain of **Prompts**. The step of the **Waterfall** being offset from the prompting behavior in the sense that you start a **Prompt** from one step and then receive the response in the following step. The progress through the series of Prompts reflects a series of _exchanges_ question/answer style with the user.

Using the **Send** function, you are quite free to send a message to the user outside of any **Prompt**. However you are doing that outside of the notion of an _exchange_ with the user, specifically, you are not waiting on a response from the user, that is you are outside of the notion of a series of **Prompts**. So in this case the progress through the **Waterfall** is designed to be explicit precisely because this behavior is not the typical behavior. Its absolutely possible to have a **Waterfall** cascade from one step to the next but its an odd thing to do so the framework makes you work for that: use the **next** function. 

Consider this: you could for example, have a **Send** call followed by another **Send** and then followed by a **Prompt**. The **Prompt** is always the last line of the implementation of a **Waterfall** step. This type system is helping you to remember that that's the typical structure. The typing is exactly the same in the C# implementation and in the JavaScript the idea is no different either, you're just not getting the compile warning.

Hopefully that makes sense.

**Comment by brynetwork** [comment]
> 
> 
> Another example of an error in TypeScript when doing a `return step.context.sendActivity('...')`
> 
> ![captureconvo](https://user-images.githubusercontent.com/4392252/52656406-249c0200-2eab-11e9-9d45-71052376363d.PNG)
> 
> Still am unable to successfully `return step.context.sendActivity('...')` in typescript

Hi TobiahZ, are you able to combine the QnA Maker inside a waterfall dialog?

I'm having an error when I put the QnAMaker.getAnswers inside a waterfall dialog,  here's the error:
![error](https://user-images.githubusercontent.com/11970029/59730396-ccb84200-9274-11e9-8b37-93fc4783a114.PNG)

Is there someone already done this before?

Thanks in advance

**Comment by TobiahZ** [comment]
@brynetwork That question is both off topic to this discussion and also doesn't contain nearly enough information. You can definitely use QnA Maker inside of a waterfall, but why you have a type error could be any number of reasons.
